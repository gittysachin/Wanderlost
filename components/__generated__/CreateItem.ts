/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: CreateItem
// ====================================================

export interface CreateItem_createItem {
  __typename: 'Item'
  id: string
}

export interface CreateItem {
  createItem: CreateItem_createItem
}

export interface CreateItemVariables {
  title: string
  description: string
  price: number
  image?: string | null
  largeImage?: string | null
}
