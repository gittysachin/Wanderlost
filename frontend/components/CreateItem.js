import React, { useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';
import Router from 'next/router';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

const uploadFile = async (e) => {
  console.log('Uploading file');
  const { files } = e.target;
  const data = new FormData();
  data.append('file', files[0]);
  data.append('upload_preset', 'sickfits');

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/davyhausser/image/upload',
    {
      method: 'POST',
      body: data,
    },
  );
  const file = await res.json();
  console.log(file);
  this.setState({
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  });
};

function CreateItem() {
  const [title, setTitle] = useState('Cool shoes');
  const [description, setDescription] = useState('Cool shoes');
  const [image] = useState('dog.jpg');
  const [largeImage] = useState('large-dog.jpg');
  const [price, setPrice] = useState(1000);

  const [createItem, { loading, error }] = useMutation(CREATE_ITEM_MUTATION, {
    variables: {
      title, description, image, largeImage, price,
    },
  });

  return (
    <Form
      onSubmit={async (e) => {
        // Stop the form from submitting
        e.preventDefault();
        // Call the mutation
        const res = await createItem();
        // Change them to the single item page
        console.log(res);
        Router.push({
          pathname: '/item',
          query: { id: res.data.createItem.id },
        });
      }}
    >
      <Error error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="file"
            name="file"
            placeholder="Upload an image"
            required
            onChange={uploadFile}
          />
          {image && (
            <img
              src={image}
              width="200"
              alt="Upload Preview"
            />
          )}
        </label>
        <label htmlFor="title">
          Title
          <input
            type="text"
            id="title"
            name="title"
            placeholder="Title"
            required
            value={title}
            onChange={(e) => setTitle(e.target.value)}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            required
            value={price}
            onChange={(e) => setPrice(formatMoney(parseFloat(e.target.value)))}
          />
        </label>
        <label htmlFor="Description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Enter A Description"
            required
            value={description}
            onChange={(e) => setDescription(e.target.value)}
          />
        </label>
        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  );
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
