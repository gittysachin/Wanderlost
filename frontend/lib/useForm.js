import { useState } from 'react';

const uploadFile = async ({ value, inputs, updateInputs }) => {
  console.log('Uploading file');
  const data = new FormData();
  data.append('file', value);
  data.append('upload_preset', 'sickfits');

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/davyhausser/image/upload',
    {
      method: 'POST',
      body: data,
    },
  );
  const file = await res.json();
  console.log(file);
  updateInputs({
    ...inputs,
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  })
};

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = e.target.files;
      uploadFile({ value, inputs, updateInputs });
    }
    updateInputs({
      ...inputs,
      [name]: e.target.value,
    });
  }

  function resetForm() {
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm
  };
}