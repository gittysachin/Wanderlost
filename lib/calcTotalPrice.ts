import { GetCurrentUser_user_cart } from '../components/__generated__/GetCurrentUser'

export default function calcTotalPrice(
  cart: Array<GetCurrentUser_user_cart>
): number {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally
    return tally + cartItem.quantity * cartItem.item.price
  }, 0)
}
