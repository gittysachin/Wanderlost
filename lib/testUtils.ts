import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'
import { GetItem_item } from '../components/__generated__/GetItem'
import { GetOrder_order, GetOrder_order_items } from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides: Partial<GetCurrentUser_user_cart>): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export { LocalStorageMock, fakeItem, fakeUser, fakeCartItem, fakeOrder, fakeOrderItem }
