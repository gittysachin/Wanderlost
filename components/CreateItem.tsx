import { useMutation, gql } from '@apollo/client'
import { useRouter } from 'next/router'
import useForm from '../lib/useForm'
import Form from './styles/Form'
import Error from './ErrorMessage'
import { ALL_ITEMS_QUERY } from './Items'
import { PAGINATION_QUERY } from './Pagination'
import { CreateItem, CreateItemVariables } from './__generated__/CreateItem'

const CREATE_ITEM_MUTATION = gql`
  mutation CreateItem(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

function CreateItem() {
  const { inputs, handleChange } = useForm()
  const router = useRouter()

  const [createItem, { loading, error }] = useMutation<CreateItem, CreateItemVariables>(
    CREATE_ITEM_MUTATION,
    {
      variables: {
        title: inputs.title,
        description: inputs.description,
        // TODO: correct type to number instead of string in useForm hook
        price: Number(inputs.price),
        image: inputs.image,
        largeImage: inputs.largeImage,
      },
      refetchQueries: [{ query: ALL_ITEMS_QUERY }, { query: PAGINATION_QUERY }],
    }
  )

  return (
    <Form
      onSubmit={async (e) => {
        e.preventDefault()
        console.log(inputs)
        const res = await createItem()
        router.push('/item/[id]', `/item/${res.data.createItem.id}`)
      }}
    >
      <Error error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="file"
            name="file"
            placeholder="Upload an image"
            required
            onChange={handleChange}
          />
          {inputs.image && <img src={inputs.image} width="200" alt="Upload Preview" />}
        </label>

        <label htmlFor="title">
          Title
          <input
            type="text"
            id="title"
            name="title"
            placeholder="Title"
            required
            value={inputs.title}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            required
            value={inputs.price}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Enter A Description"
            required
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  )
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
