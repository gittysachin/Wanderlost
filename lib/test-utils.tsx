import React from 'react'
import { render as defaultRender, RenderResult } from '@testing-library/react'
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import { NextRouter } from 'next/router'
import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'

export * from '@testing-library/react'

import { GetItem_item } from '../components/__generated__/GetItem'
import {
  GetOrder_order,
  GetOrder_order_items,
} from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// --------------------------------------------------
// Override the default test render with our own
//
// You can override the router mock like this:
//
// const { baseElement } = render(<MyComponent />, {
//   router: { pathname: '/my-custom-pathname' },
// });
// --------------------------------------------------
type DefaultParams = Parameters<typeof defaultRender>
type RenderUI = DefaultParams[0]
type RenderOptions = DefaultParams[1] & { router?: Partial<NextRouter> }

export function render(
  ui: RenderUI,
  { wrapper, router, ...options }: RenderOptions = {}
): RenderResult {
  if (!wrapper) {
    wrapper = ({ children }) => {
      const Wrapper = (
        <RouterContext.Provider value={{ ...mockRouter, ...router }}>
          {children}
        </RouterContext.Provider>
      )
      return Wrapper
    }
  }

  return defaultRender(ui, { wrapper, ...options })
}

const mockRouter: NextRouter = {
  basePath: '',
  pathname: '/',
  route: '/',
  asPath: '/',
  query: {},
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
}

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (
  overrides: Partial<GetCurrentUser_user_cart>
): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
