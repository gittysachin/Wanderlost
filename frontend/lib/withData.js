import withApollo from 'next-with-apollo';
import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
      request: (operation) => {
        operation.setContext({
          fetchOptions: {
            credentials: 'include',
          },
          headers,
        });
      },
    }),
    resolvers: {
      Mutation: {
        toggleCart(_, variables, { cache }) {
          // read the cartOpen value from the cache
          const { cartOpen } = cache.readQuery({
            query: LOCAL_STATE_QUERY,
          });
          // Write the cart state to the opposite
          const data = {
            data: { cartOpen: !cartOpen },
          };
          cache.writeData(data);
          return data;
        },
      },
    },
  });
}

export default withApollo(createClient);
