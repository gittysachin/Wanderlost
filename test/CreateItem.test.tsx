import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { MockedProvider } from '@apollo/react-testing'
import { useRouter } from 'next/router'
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem'
import { fakeItem } from '../lib/testUtils'

jest.mock('next/router', () => ({
  push: jest.fn(),
}))

const item = fakeItem()
describe('<CreateItem/>', () => {
  it('renders and matches snapshot', async () => {
    const { container } = render(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('handles state updating', async () => {
    render(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    )

    await userEvent.type(screen.getByPlaceholderText('Name'), item.title)
    await userEvent.type(screen.getByPlaceholderText('Price'), item.price.toString())
    await userEvent.type(screen.getByPlaceholderText('Description'), item.description)

    await waitFor(() => {
      expect(screen.getByDisplayValue(item.title)).toBeInTheDocument()
      expect(screen.getByDisplayValue(item.price.toString())).toBeInTheDocument()
      expect(screen.getByDisplayValue(item.description)).toBeInTheDocument()
    })
  })
  it('creates an item when the form is submitted', async () => {
    const router = useRouter()
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            description: item.description,
            image: '',
            price: item.price,
          },
        },
        result: {
          data: {
            createItem: {
              ...item,
              id: 'abc123',
              __typename: 'Item',
            },
          },
        },
      },
    ]

    render(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    )
    await userEvent.type(screen.getByPlaceholderText('Name'), item.title)
    await userEvent.type(screen.getByPlaceholderText('Price'), item.price.toString())
    await userEvent.type(screen.getByPlaceholderText('Description'), item.description)
    // mock the router
    await userEvent.click(screen.getByText('Submit'))
    await waitFor(() => {
      expect(router.push).toHaveBeenCalled()
      expect(router.push).toHaveBeenCalledWith({
        pathname: '/item',
        query: { id: 'abc123' },
      })
    })
  })
})
