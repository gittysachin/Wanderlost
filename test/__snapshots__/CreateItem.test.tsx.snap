// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> renders and matches snapshot 1`] = `
<div>
  <form
    class="sc-bdnylx dgmYIl"
  >
    <fieldset
      aria-busy="false"
    >
      <label>
        Image
      </label>
      <input
        name="file"
        type="file"
      />
      
      <label>
        Title
      </label>
      <input
        name="title"
      />
      <label>
        Price
      </label>
      <input
        name="price"
        type="number"
      />
      <label>
        Description
      </label>
      <textarea
        name="description"
      />
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</div>
`;
